lightning/restore/restore_test.go

```diff
@@ -76,3 +96,689 @@ func (s *restoreSuite) TestNewTableRestoreFailure(c *C) {
+
+	engine, err := importer.OpenEngine(ctx, "`db`.`table`", -1)
+	c.Assert(err, IsNil)
+
+	err = writeToEngine(ctx, engine, []kvenc.KvPair{})
+	c.Assert(err, IsNil)
+}
+
+func (s *restoreSuite) TestWriteToEngine(c *C) {
+	controller := gomock.NewController(c)
+	defer controller.Finish()
+	mockClient := mock.NewMockImportKVClient(controller)
+	importer := kv.NewMockImporter(mockClient, "127.0.0.1:2379")
+	mockWriteStream := mock.NewMockImportKV_WriteEngineClient(controller)
+
+	ctx := context.Background()
+
+	// The expected API call to the importer
+
+	mockClient.EXPECT().OpenEngine(ctx, gomock.Any()).Return(nil, nil)
+	mockClient.EXPECT().WriteEngine(ctx, gomock.Any()).Return(mockWriteStream, nil)
+	mockWriteStream.EXPECT().
+		Send(gomock.Any()).
+		DoAndReturn(func(req *import_kvpb.WriteEngineRequest) error {
+			c.Assert(req.GetHead(), NotNil)
+			return nil
+		})
+	mockWriteStream.EXPECT().
+		Send(gomock.Any()).
+		DoAndReturn(func(req *import_kvpb.WriteEngineRequest) error {
+			batch := req.GetBatch()
+			c.Assert(batch, NotNil)
+			c.Assert(batch.Mutations, DeepEquals, []*import_kvpb.Mutation{
+				{
+					Op:    import_kvpb.Mutation_Put,
+					Key:   []byte("k1"),
+					Value: []byte("v1"),
+				},
+				{
+					Op:    import_kvpb.Mutation_Put,
+					Key:   []byte("k2"),
+					Value: []byte("v2"),
+				},
+			})
+			return nil
+		})
+	mockWriteStream.EXPECT().CloseAndRecv().Return(nil, nil)
+
+	// Actually run the test function
+
+	engine, err := importer.OpenEngine(ctx, "`db`.`table`", -1)
+	c.Assert(err, IsNil)
+
+	err = writeToEngine(ctx, engine, []kvenc.KvPair{
+		{Key: []byte("k1"), Val: []byte("v1")},
+		{Key: []byte("k2"), Val: []byte("v2")},
+	})
+	c.Assert(err, IsNil)
+}
+
+var _ = Suite(&tableRestoreSuite{})
+
+type tableRestoreSuite struct {
+	tr  *TableRestore
+	cfg *config.Config
+}
+
+func (s *tableRestoreSuite) SetUpTest(c *C) {
+	// Produce a mock table info
+
+	p := parser.New()
+	p.SetSQLMode(mysql.ModeANSIQuotes)
+	se := tmock.NewContext()
+	node, err := p.ParseOneStmt(`
+		CREATE TABLE "table" (
+			a INT,
+			b INT,
+			c INT,
+			KEY (b)
+		)
+	`, "", "")
+	c.Assert(err, IsNil)
+	core, err := ddl.MockTableInfo(se, node.(*ast.CreateTableStmt), 0xabcdef)
+	c.Assert(err, IsNil)
+	core.State = model.StatePublic
+
+	tableInfo := &TidbTableInfo{Name: "table", core: core}
+	dbInfo := &TidbDBInfo{
+		Name:   "db",
+		Tables: map[string]*TidbTableInfo{"table": tableInfo},
+	}
+
+	// Write some sample SQL dump
+
+	fakeDataDir := c.MkDir()
+	fakeDataFilesCount := 6
+	fakeDataFilesContent := []byte("INSERT INTO `table` VALUES (1, 2, 3);")
+	c.Assert(len(fakeDataFilesContent), Equals, 37)
+	fakeDataFiles := make([]string, 0, fakeDataFilesCount)
+	for i := 1; i <= fakeDataFilesCount; i++ {
+		fakeDataPath := path.Join(fakeDataDir, fmt.Sprintf("db.table.%d.sql", i))
+		err = ioutil.WriteFile(fakeDataPath, fakeDataFilesContent, 0644)
+		c.Assert(err, IsNil)
+		fakeDataFiles = append(fakeDataFiles, fakeDataPath)
```

Comment:
All files should be deleted in `TearDownTest`.

Reply:
I'm going to move these into `SetUpSuite` since the data source is going to be immutable anyway.

Reply:
Also, `c.MkDir()` will return a [fresh directory](https://github.com/pingcap/check/blob/67f458068fc864dabf17e38d4d337f28430d13ed/check.go#L159) for every call, so the stale data from the previous tests won't affect the next ones anyway.